### PR TITLE
🍒[5.10][Distributed] Allow overloads with different types; mangling can handle it

### PR DESCRIPTION
**Pull request by ktoso**
**Description:** We too aggressively prevented overloads; The code's intention was to prevent overloads ONLY on the effects of a function (async / throws), since it complicates the relationship with thunks. Overloads ARE allowed and work just fine, so this patch fixes the too naive type system checking.
**Risk:** Low, loosens up the too restrictive checks made on distributed funcs. The mangling scheme can handle these with no problem.
**Reward:** Adopters are able to write APIs in the style they're used to, e.g. "`distributed func received(SomeEvent)`" and "`distributed func received(OtherEvent)`".
**Impact:** Small; Only distributed function overloads are affected.
**Review by:** @xedin @slavapestov 
**Testing:** CI testing; added tests to cover the situation.
**Original PR:** https://github.com/apple/swift/pull/69595
**Radar:** rdar://117818281

### CODE REVIEW

File: test/Distributed/distributed_func_overloads.swift [hunk 1]
```diff
@@ -16,25 +16,34 @@ distributed actor Overloader {
...
   distributed func overloadedDistAsync() async {}
-  // expected-note@-1{{ambiguous distributed func 'overloadedDistAsync()' declared here}}
-  // expected-note@-2{{'overloadedDistAsync()' previously declared here}}
+  // expected-note@-1{{'overloadedDistAsync()' previously declared here}}
+  // expected-error@-2{{ambiguous distributed func declaration 'overloadedDistAsync()', cannot overload distributed methods on effect only}}
   distributed func overloadedDistAsync() async throws {}
-  // expected-error@-1{{ambiguous distributed func declaration 'overloadedDistAsync()', cannot overload distributed methods on effect only}}
-  // expected-error@-2{{invalid redeclaration of 'overloadedDistAsync()'}}
+  // expected-error@-1{{invalid redeclaration of 'overloadedDistAsync()'}}
+  // expected-note@-2{{ambiguous distributed func 'overloadedDistAsync()' declared here}}
```
Comment (ktoso) [inline test/Distributed/distributed_func_overloads.swift:38]: These are just message reordering; we still diagnose all bad situations.

File: lib/Sema/TypeCheckDistributed.cpp [hunk 2]
```diff
@@ -761,19 +761,34 @@ void TypeChecker::checkDistributedActor(SourceFile *SF, NominalTypeDecl *nominal
...
+                    allSame && // diagnose if ambiguous
+                    !diagnosedAmbiguity.contains(func->getName())) {
+                  candidate->diagnose(
+                      diag::distributed_func_cannot_overload_on_async_only,
+                      candidate->getName());
+                  diagnosedAmbiguity.insert(func->getName());
+                } else if (diagnosedAmbiguity.contains(func->getName())) {
+                  candidate->diagnose(
+                      diag::distributed_func_other_ambiguous_overload_here,
+                      candidate->getName());
```
Comment (ktoso) [inline lib/Sema/TypeCheckDistributed.cpp:788]: put a note on all other members

**Comment by ktoso** [comment]
@swift-ci please test

**Comment by ktoso** [comment]
@swift-ci please test

**Comment by ktoso** [comment]
@swift-ci please test

File: lib/AST/Decl.cpp [hunk 3]
```diff
@@ -3072,9 +3072,37 @@ bool swift::conflicting(const OverloadSignature& sig1,
...
+  // are simpler.
+  //
+  // The restriction on overloads is not a problem for distributed calls,
+  // as we don't have a vast swab of APIs which must compatibly get async
+  // versions, as that is what the async overloading aimed to address.
+  //
+  // Note also, that overloading on throws is already illegal anyway.
+  if (sig1.IsDistributed || sig2.IsDistributed) {
+    if (sig1.IsAsyncFunction != sig2.IsAsyncFunction)
+      return true;
```
Comment (DougGregor) [inline lib/AST/Decl.cpp:3100]: I do not believe this is correct; if one of them is distributed, and mismatch on `async`, we should fall through to check the names below, not determine that they are always conflicting. They might have different argument names entirely.

Comment (ktoso) [reply lib/AST/Decl.cpp:3100]: Thank you, I’ll check and fix that

Comment (xedin) [reply lib/AST/Decl.cpp:3100]: Oh it’s too early, I should have took a closer look, sorry!

Comment (ktoso) [reply lib/AST/Decl.cpp:3100]: Thank you both for reviewing! 🙏

Comment (ktoso) [reply lib/AST/Decl.cpp]: https://github.com/apple/swift/pull/69688

**Comment by ktoso** [comment]
@swift-ci please test

**Comment by ktoso** [comment]
@swift-ci please test Windows

**Review by DougGregor — APPROVED** [review]

**Pull request merged**